File: AI/middleware/clarification.py
"""
第七层: 澄清拦截 (ClarificationMiddleware)

问题场景:
- 智能体目标模糊或不确定
- 行动意图不清晰

解决方案:
- 在目标模糊时拦截并请求澄清
"""

import re
from typing import Any

from .base import BaseMiddleware, MiddlewareContext, MiddlewareResult


class ClarificationMiddleware(BaseMiddleware):
    """澄清拦截中间件"""

    def __init__(self):
        super().__init__("ClarificationMiddleware")

    # Vague goal patterns
    VAGUE_PATTERNS = [
        re.compile(r'.*(?:可能|也许|大概|或者).*(?:吧|呢|吗)'),
        re.compile(r'^(?:嗯|啊|哦|呃|这个)'),
        re.compile(r'^(?:不知道|不确定|没想好)'),
        re.compile(r'^(?:随便|都可以|无所谓)'),
    ]

    def process(self, context: MiddlewareContext, agent: Any) -> MiddlewareResult:
        if not self.enabled:
            return MiddlewareResult()

        goal = context.goal or ""
        action = context.action or ""

        if context.phase.value == "before_think" and goal:
            if self._is_vague(goal):
                return MiddlewareResult(
                    should_continue=True,
                    modified_goal="explore",
                    message=f"[{self.name}] 目标模糊,建议探索: {goal}"
                )

        if context.phase.value == "before_action" and action:
            if self._is_vague(action):
                return MiddlewareResult(
                    should_continue=True,
                    message=f"[{self.name}] 行动意图不清晰: {action}"
                )

        return MiddlewareResult()

    def _is_vague(self, text: str) -> bool:
        for pattern in self.VAGUE_PATTERNS:
            if pattern.match(text):
                return True
        return False

File: src/AI/middleware/clarification.ts
/**
 * 澄清中间件
 * 用于检测和处理需要用户澄清的情况
 */
import { BaseMiddleware, MiddlewareContext, MiddlewareResult } from './base';

export class ClarificationMiddleware extends BaseMiddleware {
  /**
   * 构造函数
   */
  constructor() {
    super("ClarificationMiddleware");
  }

  /**
   * 处理方法
   * 检查是否需要用户澄清
   * @param context 中间件上下文
   * @param agent 智能体实例
   * @returns 中间件结果
   */
  process(context: MiddlewareContext, agent: any): MiddlewareResult {
    // 检查是否需要澄清
    // 这里可以实现需要澄清的逻辑，例如：
    // 1. 检测用户输入是否模糊
    // 2. 检测是否缺少必要的信息
    // 3. 检测是否存在歧义
    
    // 目前是一个空实现，返回继续执行
    return {
      should_continue: true,
      metadata: {}
    };
  }
}
